Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm

```diff
@@ -35,60 +37,58 @@
  * document key in memory. This is only an estimate and includes the size
  * of the segments of the path, but not any object overhead or path separators.
  */
-static size_t FSTDocumentKeyByteSize(FSTDocumentKey *key) {
+static size_t FSTDocumentKeyByteSize(const DocumentKey &key) {
   size_t count = 0;
-  for (const auto &segment : key.path) {
+  for (const auto &segment : key.path()) {
     count += segment.size();
   }
   return count;
 }
 
 @interface FSTMemoryRemoteDocumentCache ()
 
-/** Underlying cache of documents. */
-@property(nonatomic, strong) FSTMaybeDocumentDictionary *docs;
-
 @end
 
-@implementation FSTMemoryRemoteDocumentCache
-
-- (instancetype)init {
-  if (self = [super init]) {
-    _docs = [FSTMaybeDocumentDictionary maybeDocumentDictionary];
-  }
-  return self;
+@implementation FSTMemoryRemoteDocumentCache {
+  /** Underlying cache of documents. */
+  MaybeDocumentMap _docs;
 }
 
 - (void)addEntry:(FSTMaybeDocument *)document {
-  self.docs = [self.docs dictionaryBySettingObject:document forKey:document.key];
+  _docs = _docs.insert(document.key, document);
 }
 
 - (void)removeEntryForKey:(const DocumentKey &)key {
-  self.docs = [self.docs dictionaryByRemovingObjectForKey:key];
+  _docs = _docs.erase(key);
 }
 
 - (nullable FSTMaybeDocument *)entryForKey:(const DocumentKey &)key {
-  return self.docs[static_cast<FSTDocumentKey *>(key)];
+  auto found = _docs.find(key);
+  return found != _docs.end() ? found->second : nil;
 }
 
-- (FSTDocumentDictionary *)documentsMatchingQuery:(FSTQuery *)query {
-  FSTDocumentDictionary *result = [FSTDocumentDictionary documentDictionary];
+- (MaybeDocumentMap)documentsMatchingQuery:(FSTQuery *)query {
+  MaybeDocumentMap result;
 
   // Documents are ordered by key, so we can use a prefix scan to narrow down the documents
   // we need to match the query against.
-  FSTDocumentKey *prefix = [FSTDocumentKey keyWithPath:query.path.Append("")];
-  NSEnumerator<FSTDocumentKey *> *enumerator = [self.docs keyEnumeratorFrom:prefix];
-  for (FSTDocumentKey *key in enumerator) {
-    if (!query.path.IsPrefixOf(key.path)) {
+  DocumentKey prefix{query.path.Append("")};
+  for (auto it = _docs.lower_bound(prefix); it != _docs.end(); ++it) {
+    const DocumentKey &key = it->first;
+    if (!query.path.IsPrefixOf(key.path())) {
       break;
     }
-    FSTMaybeDocument *maybeDoc = self.docs[key];
+    FSTMaybeDocument *maybeDoc = nil;
+    auto found = _docs.find(key);
+    if (found != _docs.end()) {
+      maybeDoc = found->second;
+    }
     if (![maybeDoc isKindOfClass:[FSTDocument class]]) {
       continue;
     }
-    FSTDocument *doc = (FSTDocument *)maybeDoc;
+    FSTDocument *doc = static_cast<FSTDocument *>(maybeDoc);
```

Comment:
(again)

Reply:
Oh; and to be clear, this usage here is perfectly fine due to the isKindOfClass check immediately preceding. Plus it being exactly analogous to the previous version. Don't know why I commented on this usage in the first place (probably wan't paying enough attention.)